spectator-web/src/main/java/com/netflix/spectator/controllers/MetricsController.java

```diff
@@ -0,0 +1,207 @@
+    TagMeasurementFilter queryFilter = new TagMeasurementFilter(
+        filterMeterNameRegex, filterTagNameRegex, filterTagValueRegex);
+    MeasurementFilter filter;
+    if (all) {
+      filter = queryFilter;
+    } else {
+      filter = new ChainedMeasurementFilter(queryFilter,
+                                            getDefaultMeasurementFilter());
+    }
+
+    ApplicationRegistry response = new ApplicationRegistry();
+    response.applicationName = applicationName;
+    response.metrics = encodeRegistry(registry, filter);
+    return response;
+  }
+
+  MetricValuesMap encodeRegistry(Registry registry, MeasurementFilter filter) {
+    MetricValuesMap metricMap = new MetricValuesMap();
+
+    /**
+     * Spectator meters seam to group measurements. The meter name is the
+     * measurement name prefix. It seems that all the measurements within a
+     * meter instance share the same tag values as the meter.
+     * Different instances are different tag assignments, keeping the groups
+     * of measurements together.
+     *
+     * For now, we are not going to make this assumption and will just flatten
+     * out the whole measurement space. That makes fewer assumptions for now.
+     * If we make grouping assumptions we can factor out the labels into a more
+     * terse json. But the difference isnt necessarily that big and if the
+     * assumption doesnt hold, then forming the groupings will be complicated
+     * and more expensive.
+     */
+    for (Meter meter : registry) {
+      Map<Id, List<Measurement>> collection
+          = new HashMap<Id, List<Measurement>>();
+      collectValues(collection, meter, filter);
+
+      if (!collection.isEmpty()) {
+        String kind = meterToKind(meter);
+        for (Map.Entry<Id, List<Measurement>> entry : collection.entrySet()) {
+           String entryName = entry.getKey().name();
+           MetricValues have = metricMap.get(entryName);
+           if (have == null) {
+             metricMap.put(entryName, MetricValues.make(kind, entry.getValue()));
+           } else {
+             have.addMeasurements(kind, entry.getValue());
+           }
+        }
+      }
+    }
+    return metricMap;
+  }
+
+  /**
+   * Collect all the meter values matching the tag pattern.
+   */
+  public static List<Id> collectValues(
+      Map<Id, List<Measurement>> collection,
+      Meter meter,
+      MeasurementFilter filter) {
+    List<Id> new_ids = new ArrayList<Id>();
+
+    for (Measurement measurement : meter.measure()) {
+      if (!filter.keep(meter, measurement)) {
+          continue;
+      }
+      Id id = measurement.id();
+
+      List<Measurement> valueList = collection.get(id);
+      if (valueList == null) {
+        valueList = new ArrayList<Measurement>();
+        collection.put(id, valueList);
+        new_ids.add(id);
+      }
+      valueList.add(measurement);
+    }
+    return new_ids;
+  }
+
+  public static String meterToKind(Meter meter) {
```

Comment:
I'll chat with some others on our team about this and try to get #312 fixed today. We should be able to reduce some of the pain around this.